### ISSUE TITLE
Shapes layer performance

### ISSUE THREAD

**Issue by kevinyamauchi** [napari/napari#1562]
This work in progress issue is to discuss the current state of the Shapes layer performance (mostly with regards to frame rate with large numbers of shapes) and potential improvements. I am still exploring and I will update the original post as I learn more.

When creating a `Shapes` layer with many shapes (500-1000), the performance degrades substantially. In particular, the layer creation time is long and the interface is not a responsive (e.g., selecting and dragging shapes).

## Approach
Using, @pwinston 's awesome profiling tool, I profiled the startup and selection of shapes. The scripts to generate the viewer, the profiling config, and the traces are available in [this repository](https://github.com/kevinyamauchi/profile-napari-shapes)). I ran the profiling using #1550 since the qt profiling is currently broken on master.

## Responsiveness during dragging
When creating a `Shapes` layer with 1000 rectangles, selecting and dragging a shape is not very responsive. We see that the majority of the time in the `mouse_move` callback is spent updating the thumbnail (trace from `many_shapes_one_slice` example in this [this repository](https://github.com/kevinyamauchi/profile-napari-shapes)).

![Screenshot 2020-08-18 at 13 34 51](https://user-images.githubusercontent.com/1120672/90508236-9ebfdd80-e157-11ea-9ea4-d5992b1c30d3.png)

This is a fairly straightforward fix and by disabling the thumbnail update during the drag and calling it once at the end greatly improves performance (from ~280 ms down to ~33 ms for the `mouse_move` event).

![Screenshot 2020-08-18 at 13 37 47](https://user-images.githubusercontent.com/1120672/90508449-0fff9080-e158-11ea-9d04-aef17b871a32.png)

We see there is still a long time being spent performing the shape highlighting (~10 ms of the ~30 ms in the `mouse_move` event). This is likely spend in the `Shapes._outline_shapes()` method. There is also some unaccounted for time that I haven't chased up yet. This also does not account for the time that vispy spends drawing the shapes, but this issue is focusing on improvements we can make to the model.

## Slow startup
Currently, starting a viewer with a lot of shapes take a long time. In the image below, I have highlighted adding one rectangle during the initialization of a Shapes layer with 500 rectangles (~5 ms). Creating a layer with 5000 rectangles takes ~25s! Note that a lot of time is spent with repeated calls to re-index the mesh z order (once per shape). Trace from `many_shapes_many_slices` example in this [this repository](https://github.com/kevinyamauchi/profile-napari-shapes))

![Screenshot 2020-08-18 at 14 31 48](https://user-images.githubusercontent.com/1120672/90513244-98cdfa80-e15f-11ea-9edb-afb746525c96.png)

If we wait to re-index the mesh z order until after adding all of the shapes, we can cut the `Shapes.___init__` time by ~75% (from ~2.9s to ~0.7s for 5 slices, 100 shapes per slice).

![Screenshot 2020-08-18 at 14 36 03](https://user-images.githubusercontent.com/1120672/90513662-48a36800-e160-11ea-80fd-58bac5638915.png)

We see that the majority of the time spent creating each shape is now the mesh generation. The shapes used in this example are rectangles, so shapes with more complex geometries will likely take longer to mesh.

Also, if we "zoom out" a bit, we see that there are repeated calls to generate the thumbnail during initialization. Eliminating those will also save time.

![Screenshot 2020-08-18 at 14 48 52](https://user-images.githubusercontent.com/1120672/90514805-f4998300-e161-11ea-8f43-5b73daecacc5.png)

## Responsiveness during highlighting

Currently, there is considerable lag between the mouse button release when dragging to select a bunch of shapes and when the highlights are drawn. Since the highlight drawing is blocking, the GUI is unresponsive during this time. From profiling, we see it takes ~1.5s to draw the highlights over ~300 shapes. There are two main issues: (1) we call the `set_highlight()` method twice and (2) the `ShapesList.outline()` method is slow (~750 ms for ~300 shapes) and is taking up most of the time. Therefore, I propose we make a small PR to eliminate the duplicate the highlight call (i.e., easy 2x speedup) and then investigate ways to make the generation/indexing of the outline mesh faster. 

<img width="1552" alt="Screenshot 2020-08-25 at 09 49 27" src="https://user-images.githubusercontent.com/1120672/91147296-5bfe8800-e6b8-11ea-9a48-ead127df18e2.png">

After a bit more profiling:

- There are some large for loops that can be vectorized in `ShapesList.outline()`, which should speed up the outline mesh generation quite a bit. See [here](https://github.com/napari/napari/blob/67eb6908f77d93d064a69a6b4f6b48034ac60a90/napari/layers/shapes/_shape_list.py#L670-L682)
- The calculations to determine which shapes are in the selection box are also slow (~1.5 s for layer with 1000 shapes). See [here](https://github.com/napari/napari/blob/67eb6908f77d93d064a69a6b4f6b48034ac60a90/napari/layers/shapes/_shape_list.py#L710-L732).

## Proposal
- [x] Add benchmarks for selecting and dragging Shapes (see #1563 )
- [x]  In a first PR, we make some small changes to improve responsiveness (see #1561)

    - prevent the thumbnail (improve dragging responsiveness)
    - only calculate z order once when initializing the Shapes model (reduce startup time)
- [ ] Speed up highlighting by eliminating redundant `set_highlight()` call
- [ ] Investigate speeding up generating/indexing outline meshes to speed up highlighting
- [ ] Investigate speeding up determining which shapes are in the selection box (speeds up highlighting)
- [ ] Investigate ways to speed up thumbnail generation for when we can't avoid calling `Shapes._update_thumbnail()`
- [ ] Investigate ways to speed up the meshing of the shapes (e.g., numba?)

It is probably worth noting that all of these tests were performed on my super wimpy Macbook Air. People with faster computers will likely have better performance.

**Comment by jni** [comment]
Awesome analysis @kevinyamauchi and great plan for moving forward! 👏👏👏 🎉

**Comment by sofroniewn** [comment]
Wow, this is so cool @kevinyamauchi and so great to see the perfmon work from @pwinston paying dividends here. I'm not surprised to hear there is much we can do to improve shapes performance, and some of it appears low hanging fruit!! Sounds like you're already off to the races, but we can keep discussing as you did deeper. Definitely some of those reindexing calls are costly!!

**Comment by pwinston** [comment]
Performance monitoring is fun and easy once you get into it! Really improves the quality of the conversations. I only used it for async, but I think it can be used in many ways. Just have to get creative.

**Comment by pwinston** [comment]
Using a full profiler might also bear fruit. I have not tried one on napari. A full profiler can sometimes find things that tracing cannot. I think you want tracing no matter what, but a full profiler once in a while might make sense to really exhaustively look at the cost of every single function. YMMV. Historically full profilers slow everything way down. But maybe there are good ones today for Python. Would be interesting to research it some day.

**Comment by sofroniewn** [comment]
> Using a full profiler might also bear fruit. I have not tried one on napari. 

I've had pretty good results using [airspeed velocity](https://asv.readthedocs.io/en/stable/) for profiling, we're also using for our benchmarking in general. You can see instructions here https://napari.org/docs/developers/BENCHMARKS.html#profiling, but we might @kevinyamauchi want to make some better benchmarks to profile for shapes too, this is all we have right now https://github.com/napari/napari/blob/master/napari/benchmarks/benchmark_shapes_layer.py

**Comment by kevinyamauchi** [comment]
Thanks for the feedback, everyone!

> I've had pretty good results using airspeed velocity for profiling, we're also using for our benchmarking in general. You can see instructions here https://napari.org/docs/developers/BENCHMARKS.html#profiling, but we might @kevinyamauchi want to make some better benchmarks to profile for shapes too, this is all we have right now https://github.com/napari/napari/blob/master/napari/benchmarks/benchmark_shapes_layer.py

I think it might makes sense to flesh out the Shapes benchmarks a bit before finishing up #1561 , that way we know that we are definitely improving performance. @pwinston 's perfmon tool has highlighted some key methods we should add to the benchmarking, as they are clearly bottlenecks that we should pay attention to (e.g., ~~thumbnail generation~~ and the meshing). Super cool!

edit: the update thumbnail method is already in the benchmarks 🤦

**Comment by kevinyamauchi** [comment]
Another latent step when interacting with large shapes layers is clicking on shapes to select them. From the profiling, we see that there are some duplicate calls (likely multiple refresh-triggering events being emitted) that we can clean up. However, with 1000 shapes, a single thumbnail update takes ~250 ms on my computer. Until we figure out a way to speed up the thumbnail generation, what do you all think about skipping the thumbnail update when selecting a shape in a Shapes layer with "a lot" of shapes? The updated thumbnail only shows the highlighted shape, which is likely not very visible in a layer with a lot of shapes, so I don't think it's a huge loss. Thoughts?

![Screenshot 2020-08-19 at 10 47 44](https://user-images.githubusercontent.com/1120672/90613279-6e387c00-e209-11ea-9c47-cd11a9e0f23d.png)

**Comment by kevinyamauchi** [comment]
I have made a small PR (#1563) to add asv benchmarks for selecting and dragging shapes. These should help us track how the changes we make in #1561 affect performance.

**Comment by sofroniewn** [comment]
> Until we figure out a way to speed up the thumbnail generation, what do you all think about skipping the thumbnail update when selecting a shape in a Shapes layer with "a lot" of shapes? The updated thumbnail only shows the highlighted shape, which is likely not very visible in a layer with a lot of shapes, so I don't think it's a huge loss. Thoughts?

I think fine to make thumbnail generation only happen after data editing is finished

**Comment by kevinyamauchi** [comment]
I just wanted to record a couple of ideas from the sync:
- We can add a time check to slow functions (e.g., `layer._update_thumbnail()`) to ensure they aren't called too frequently (e.g., `if time_since_last call > threshold:`)
- We can make the thumbnail generation async
- We can try to speed up functions with a lot of calculations and loops (e.g., triangulation and thumbnail generation) with numba or cython

**Comment by pwinston** [comment]
The other idea about "called too frequently" is have an explicit way for the caller suppress events/updates, consider this:

    do_something()  # this will cause the UI to update

compared to this:

    with do_not_update_the_ui():
        for i in range(1000):
            do_something()

The nice thing is when `do_not_update_the_ui()` exits it updates the UI. You can think of it as bundling or grouping many things into one update. I specifically remember doing this when working on [freeform](https://tobeva.com/about/experience/cad/) around 20 years ago!

The timing thing is good for "somehow this is getting called too often" without really caring what's causing those updates. While explicit disabling might be better if you know ahead of time you are going to spam things.

**Comment by pwinston** [comment]
I had a totally different idea. Sometimes with graphics you can create one of something and then instance it a bunch of times. I don't think that really applies here, but graphics cards highly like this approach. Instead of of drawing 100,000 different trees, you have only 10 unique trees, but you have around 10,000 instances of each tree. The speedup there is gigantic because you only send 10 tree models to the card, followed by a list of 100,000 transforms. Instead of sending 100,000 "different" tree models, where most of them are duplicates.

I remember once creating 3d arrows. But you create create the pointy end-cap just once, and use it over and over on every type of arrow, curved or straight, etc. 

Sometimes there's CPU-side versions of the idea. Can you factor out something  you are doing or storing that's really duplicated? Like if you are need to do geometry at joints, can you have some common joints pre-computed like 90 degrees? And just slap those in? Which is a type of caching but maybe done very explicitly and by hand.

It's a very general idea. Sometimes the straightforward "just do it from scratch N times" is really simple. And the factoring out and combining parts is a lot more complicated. But sometimes it's tons faster and worth the effort.

**Comment by kevinyamauchi** [comment]
> The nice thing is when do_not_update_the_ui() exits it updates the UI. You can think of it as bundling or grouping many things into one update. I specifically remember doing this when working on freeform around 20 years ago!

Good idea! I used a [similar context manager](https://github.com/napari/napari/blob/fe4d7d59dcac7832a9d444295a0567fd5b39e48d/napari/layers/base/base.py#L643-L647)) to block the thumbnail refresh when making a batch of calls to other functions that would all call the thumbnail refresh function. However, I didn't think to use that same context manager to then call the update at the end. Nice!

> Sometimes there's CPU-side versions of the idea. Can you factor out something you are doing or storing that's really duplicated? Like if you are need to do geometry at joints, can you have some common joints pre-computed like 90 degrees? And just slap those in? Which is a type of caching but maybe done very explicitly and by hand.

I like this line of thinking. After #1561, the the thumbnail update will be one of the slowest aspects of Shapes layer. We have a few different shape types (e.g., rectangle, line, polygon). Perhaps we can sort them by type and somehow either batch or use the optimal algorithm for each type. 

Alternatively, perhaps we can perform the rasterization on the mesh triangles we have precomputed instead of using the full polygon vertices as we are now. I haven't had a look, but I am guessing there are fast rasterization algorithms for triangles. At the very least, maybe we constrain our pixel search to the bounding box around the triangle (probably small number of pixels in the downsampled thumbnail).

P.S. that haptic pen for FreeForm looks awesome!

**Comment by pwinston** [comment]
Yea batching/grouping/sorting comes up a lot in graphics. Generally you want to minimize state transitions with graphics (and with CPU really). It's like "once I'm setup to do X, let's do all the X's". I'm not really sure what rasterization we are doing today. Naively I'd expect you'd want the card to everything possible. 

Like back in the day with OpenGL they had crazily detailed lines and joints it could make for you:
https://www.opengl.org/archives/resources/code/samples/sig99/advanced99/notes/node281.html

I really don't know much about what we are doing. Someday a nice walkthrough of Surface or all the geometry layers would be nice, to widen the circle of people that understand it. I've done 100% image layer stuff on napari so far, so Surface is like going to Europe, it seems like similar yet very different at the same time.

**Comment by pwinston** [comment]
> P.S. that haptic pen for FreeForm looks awesome!

It was $10,000 back twenty years ago. It's still for sale today but they say "contact us for pricing". You could write and ask if there a volume discount if you buy more than 100 seats. I bet they'd call you back very quickly.

**Comment by kevinyamauchi** [comment]
> I really don't know much about what we are doing. Someday a nice walkthrough of Surface or all the geometry layers would be nice, to widen the circle of people that understand it. I've done 100% image layer stuff on napari so far, so Surface is like going to Europe, it seems like similar stuff yet is totally different.

I totally understand where you're coming from. I think we've been trying to find time to document/communicate as much as possible while still moving things forward, but there is definitely a lot of "siloed" knowledge.

In this particular problem, we have a bunch of polygons stored as coordinates of mesh vertices that we want to convert into a thumbnail image. The current way it is done is quite slow (~300 ms for 1000 shapes). We try to keep the thumbnail up to date with what is in view in the viewer. However, with a 300 ms calculation time, we obviously have to minimize the number of updates to avoid blocking the GUI. That is the motivation for trying to speed up the generation of the thumbnail from the shape meshes.

The current algorithm loops through the shapes, generates a mask for each shape and then uses those masks to apply the color to the thumbnail image. However, it doesn't check to if the shape is covered by another shape, which when there are a lot of shapes, is probably wasteful (especially because we downsample the thumbnail).

Perhaps we should instead take an approach that I think is more similar to ray tracing. That is loop through the pixels in the thumbnail and see which shape they hit first (i.e., checking if they are inside the shape in their z order and stopping once we find a shape containing that pixel).

In any case, I am new to this type of programming and I appreciate your insights/suggestions, @pwinston !

**Comment by pwinston** [comment]
What you definitely want to do long term is do all the rendering on the card. You do a "render to texture" instead of a render to screen. Then that texture can be used in the scene if drawing it with vispy. Or if the Qt is doing the display, you can read the texture back from the card and give it to Qt.

I mentioned C++ can be 10-1000 times faster than Python. The card is probably 10,000 times faster than Python, maybe much more, at munging geometry and drawing pixels. There's really no other way to do it if you have a lot of geometry. 

A thumbnail is similar to picture-in-picture or "portal" in a game. For instance imagine in a game there's a TV screen in the room, and on the TV screen is the view from a security camera of that same room. So the player can seem themselves on the screen.

The way you do that is do a full render pass _from the point of view of the security camera_, but you render to a texture not the screen. Then you render to the scene where the TV is showing that texture (the one you rendered from the POV of security camera). Another example in games is rendering a top-down view for use as a mini-map.

Portals were featured in the game [Portal](<https://en.wikipedia.org/wiki/Portal_(video_game)>):
* [Portal Game Trailer](https://youtu.be/TluRVBhmf8w)
* [How were the portals created?](https://youtu.be/_SmPR5mvH7w)

See also this [Non-Euclidian Game Engine](https://youtu.be/kEB11PQ9Eo8) video for a weirder example.

**Comment by pwinston** [comment]
The [gap](https://www.nextplatform.com/2019/07/10/a-decade-of-accelerated-computing-augurs-well-for-gpus/) between CPU and GPU performance is growing rapidly, to say the least. It's almost always better to use the card. Unless moving the data to/from the card is really expensive, which is sometimes the case.

<img width="700" alt="cpu-gpu-gap" src="https://user-images.githubusercontent.com/4163446/90792626-b07cbe80-e2d8-11ea-970e-1f0bb80863cc.png">

Like sending a k-d tree to the card and doing a million lookups is smart. Sending a k-d tree to the card to do one lookup is probably pointless.

**Comment by pwinston** [comment]
For this specific problem though doing some small improvements if they are enough for now might make sense. But long term with lots of geometry it probably has to move to the card.

**Comment by sofroniewn** [comment]
Just catching up with this convo, but completely agree with everything that was said above.

First off, we should definitely do a deep dive @pwinston on the "geometry" or "mesh" layers, I suspect lots of low hanging fruit there.

Second, we should probably move all thumbnail calculation off of our models and onto the card, though we might want to preserve the CPU calculations somewhere. There could be a global clean up here that might improve lots of things.

**Comment by kevinyamauchi** [comment]
> First off, we should definitely do a deep dive @pwinston on the "geometry" or "mesh" layers, I suspect lots of low hanging fruit there.

Definitely!

> Second, we should probably move all thumbnail calculation off of our models and onto the card, though we might want to preserve the CPU calculations somewhere. There could be a global clean up here that might improve lots of things.

I agree with this and I also agree with @pwinston 's comment that there may be some smaller improvements we can make now to hold us over as we figure out the "best" way to render the thumbnail and do the meshing.

**Comment by pranathivemuri** [comment]
@sofroniewn @kevinyamauchi and everyone else! Thanks so much for the great discussion here, I was wondering if the shape layer was modified after the discussion here. I was going to open another issue regarding the performance of the shape layer where it was taking a long time as well to load the boxes https://github.com/czbiohub/napari-bb-annotations/blob/pranathi-tflite/napari_bb_annotations/_key_bindings.py#L378

Please let me know, thank you!

**Comment by sofroniewn** [comment]
Not really. I think it's ok to make a new issue with your use case, we can always link them. Looks like we are now capping thumbnail generation at 100 shapes, https://github.com/napari/napari/blob/609fd9f4cc54832c752d3691d7f16e15fbf86517/napari/layers/shapes/shapes.py#L298,so that's not going to cause a big scaling issue anymore but you can also block it completely using this context manager https://github.com/napari/napari/blob/master/napari/layers/shapes/shapes.py#L2195 or

**Comment by psobolewskiPhD** [comment]
The numbers in this thread are vastly out dated, with 1000 shapes no longer being slow in any of the cases cited.
Issues con't really manifest until 10s of K.
While performance issues remain the big ones are moving shapes and adding shapes:
https://github.com/napari/napari/issues/8650
https://github.com/napari/napari/issues/8399
